In [1]:
import pandas as pd
import json

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Ajustar la opción para mostrar más caracteres en cada columna
pd.set_option('display.max_colwidth', None)

In [ ]:
# Abrir el archivo JSON y leer línea por línea
with open(r'C:\\Users\\Pablo\\OneDrive\\Henry\\ProyectoFinal\\Google Maps\\metadata-sitios\\1.json') as file:
    data = [json.loads(line) for line in file]

# Convertir la lista de diccionarios en un DataFrame
df1 = pd.DataFrame(data)

In [ ]:
# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\2.json'
# Leer el JSON en un DataFrame
df2 = pd.read_json(ruta_json, lines=True)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\3.json'
# Leer el JSON en un DataFrame
df3 = pd.read_json(ruta_json, lines=True)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\4.json'
# Leer el JSON en un DataFrame
df4 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\5.json'
# Leer el JSON en un DataFrame
df5 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\6.json'
# Leer el JSON en un DataFrame
df6 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\7.json'
# Leer el JSON en un DataFrame
df7 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\8.json'
# Leer el JSON en un DataFrame
df8 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\9.json'
# Leer el JSON en un DataFrame
df9 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\10.json'
# Leer el JSON en un DataFrame
df10 = pd.read_json(ruta_json, lines=True)

In [ ]:
def extractCity(df):
    """
    Esta función toma un DataFrame que contiene una columna llamada 'address',
    la cual tiene direcciones formateadas con comas. La función extrae el valor
    que está entre la segunda y tercera coma (que representa la ciudad) y lo
    almacena en una nueva columna llamada 'city'.

    Parámetros:
    df (DataFrame): El DataFrame que contiene la columna 'address'.

    Retorna:
    DataFrame: El DataFrame original con una nueva columna 'city' que contiene
               el valor entre la segunda y tercera coma de 'address'.
    """

    # Dividir la columna 'address' en partes usando la coma como delimitador
    # Luego, extraer el segundo elemento (índice 2), que está entre la segunda y tercera coma

    df['city'] = df['address'].str.split(',').str[2].str.strip()

    return df

In [ ]:
def nulosYVacios(df):
    """
    Este procedimiento recibe un DataFrame y muestra la cantidad de valores nulos, vacíos y la cantidad de
    valores duplicados en todas las columnas excepto en las 6, 10, 11 y 13 del DataFrame. También muestra
    el mínimo, máximo y media para columnas de tipo float e integer.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    print("Informe de Nulos, Vacíos y Duplicados:")
    print("-" * 30)

    # Total de registros en el DataFrame
    total_registros = len(df)

    # Imprimir total de registros
    print(f"Total de registros: {total_registros}")
    print("-" * 30)

    # Iterar sobre todas las columnas para mostrar nulos y vacíos
    for columna in df.columns:
        # Contar valores nulos
        nulos = df[columna].isnull().sum()

        # Contar valores vacíos (considerando que solo se cuentan si son cadenas)
        vacios = (df[columna] == '').sum() if df[columna].dtype == 'object' else 0

        # Calcular porcentajes
        porcentaje_nulos = (nulos / total_registros) * 100 if total_registros > 0 else 0
        porcentaje_vacios = (vacios / total_registros) * 100 if total_registros > 0 else 0

        # Imprimir resultados
        print(f"Columna: {columna}")
        print(f"  Valores Nulos: {nulos}")
        print(f"  Valores Vacíos: {vacios}")
        print(f"  Porcentaje Nulos: {porcentaje_nulos:.2f}%")
        print(f"  Porcentaje Vacíos: {porcentaje_vacios:.2f}%")

        # Contar y mostrar la cantidad de valores duplicados solo si no es una de las columnas con datos anidados
        if columna not in [df.columns[6], df.columns[10], df.columns[11], df.columns[13]]:
            # Filtrar los valores no nulos antes de contar duplicados
            cantidad_duplicados = df[columna].dropna().duplicated(keep=False).sum()
            print(f"  Cantidad de valores duplicados (solo no nulos): {cantidad_duplicados}")

        # Mostrar mínimo, máximo y media si la columna es de tipo float o int
        if df[columna].dtype in ['float64', 'int64']:
            minimo = df[columna].min()
            maximo = df[columna].max()
            media = df[columna].mean()
            print(f"  Mínimo: {minimo}")
            print(f"  Máximo: {maximo}")
            print(f"  Media: {media:.2f}")

        print("-" * 30)


Análisis preliminar de calidad de datos

In [ ]:
# Obtener la ciudad donde se encuentra el sitio con reseña
extractCity(df1)
extractCity(df2)
extractCity(df3)
extractCity(df4)
extractCity(df5)
extractCity(df6)
extractCity(df7)
extractCity(df8)
extractCity(df9)
extractCity(df10)

In [ ]:
def rankingFrecuenciaCiudad(df, x=10):
    """
    Este procedimiento recibe un DataFrame y un número entero x.
    Muestra en pantalla el ranking de los primeros x elementos con mayor frecuencia en la columna 'city'.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    x (int): El número de elementos a mostrar en el ranking.
    """
    print(f"Ranking de los primeros {x} elementos más frecuentes en la columna 'city':")
    print("-" * 60)

    # Verifica si la columna 'city' existe en el DataFrame
    if 'city' in df.columns:
        # Obtener la frecuencia de cada valor en la columna 'city'
        frecuencias = df['city'].value_counts()

        # Obtener los primeros x elementos con mayor frecuencia
        ranking = frecuencias.head(x)

        # Mostrar el ranking
        for index, value in ranking.items():
            print(f"{index}: {value}")
    else:
        print("La columna 'city' no existe en el DataFrame.")

    print("-" * 60)

In [ ]:
def mostrarCantidadFilasDuplicadas(df):
    """
    Filtra las filas duplicadas en un DataFrame según columnas específicas,
    y muestra la cantidad de filas duplicadas en la salida de pantalla.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    # Filtrar duplicados considerando las columnas específicas
    filas_duplicadas = df[df.duplicated(subset=[df.columns[i] for i in [0, 1, 2, 4, 5]], keep=False)]

    # Contar la cantidad de filas duplicadas
    cantidad_duplicados = len(filas_duplicadas)

    # Mostrar la cantidad de filas duplicadas
    print(f"Cantidad de filas duplicadas: {cantidad_duplicados}")

Frecuencia de recomendaciones por ciudad segun google Maps

In [ ]:
# Obtener la frecuencia de cada valor en la columna 'city'
rankingFrecuenciaCiudad(df1)

Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2599
New York: 1903
Chicago: 1799
Los Angeles: 1663
Brooklyn: 1550
San Antonio: 1200
Dallas: 1154
Miami: 1139
Las Vegas: 1089
Philadelphia: 1044
------------------------------------------------------------


In [ ]:
rankingFrecuenciaCiudad(df2)

In [ ]:
rankingFrecuenciaCiudad(df3)

In [ ]:
rankingFrecuenciaCiudad(df4)
rankingFrecuenciaCiudad(df5)
rankingFrecuenciaCiudad(df6)
rankingFrecuenciaCiudad(df7)
rankingFrecuenciaCiudad(df8)
rankingFrecuenciaCiudad(df9)
rankingFrecuenciaCiudad(df10)

Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2511
New York: 1855
Los Angeles: 1713
Chicago: 1679
Brooklyn: 1596
Miami: 1337
San Antonio: 1249
Las Vegas: 1229
Philadelphia: 1204
Dallas: 1143
------------------------------------------------------------
Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2504
New York: 1825
Chicago: 1763
Los Angeles: 1625
Brooklyn: 1610
San Antonio: 1299
Dallas: 1185
Miami: 1148
Phoenix: 1071
Philadelphia: 1051
------------------------------------------------------------
Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 2598
Chicago: 1812
New York: 1810
Brooklyn: 1636
Los Angeles: 1401
San Antonio: 1366
Dallas: 1288
Miami: 1232
Las Vegas: 1172
Phoenix: 1119
------------------

In [ ]:
nulosYVacios(df1)

Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 275001
------------------------------
Columna: name
  Valores Nulos: 7
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 84218
------------------------------
Columna: address
  Valores Nulos: 10062
  Valores Vacíos: 0
  Porcentaje Nulos: 3.66%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 51265
------------------------------
Columna: gmap_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 53146
------------------------------
Columna: description
  Valores Nulos: 261846
  Valores Vacíos: 0
  Porcentaje Nulos: 95.22%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 10123
------------------------------
Columna: latitude
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 275001 entries, 0 to 275000
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   name              274994 non-null  object 
 1   address           264939 non-null  object 
 2   gmap_id           275001 non-null  object 
 3   description       13155 non-null   object 
 4   latitude          275001 non-null  float64
 5   longitude         275001 non-null  float64
 6   category          272740 non-null  object 
 7   avg_rating        275001 non-null  float64
 8   num_of_reviews    275001 non-null  int64  
 9   price             13450 non-null   object 
 10  hours             192448 non-null  object 
 11  MISC              194972 non-null  object 
 12  state             195523 non-null  object 
 13  relative_results  238771 non-null  object 
 14  url               275001 non-null  object 
 15  city              261516 non-null  object 
dtypes: float64(3), int64

Puede ser que se encuentren valores repetidos de lugar en el mapa, porque quizas es un edificio, o antes estuvo otro, por lo tanto habra que ver si es que tienen el mismo nombre. coincidencia entre esas coolumnas.
 0   name              274994 non-null  object
 1   address           264939 non-null  object
 2   gmap_id           275001 non-null  object
 4   latitude          275001 non-null  float64
 5   longitude         275001 non-null  float64
 14  url               275001 non-null  object

In [ ]:
# Filas duplicas considerando las columnas 0 a 5
mostrarCantidadFilasDuplicadas(df1)

Cantidad de filas duplicadas: 53146


In [ ]:
# Cargar el archivo de business
dfYeplp = pd.read_pickle(r'C:\Users\Pablo\OneDrive\Henry\Proyecto Final\Yelp\business.pkl')

In [ ]:
dfYeplp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 28 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
 14  business_id   5 non-null       object
 15  name          5 non-null       object
 16  address       5 non-null       object
 17  city          5 non-null       object
 18  state         5 non-null     

In [ ]:
# Filtrar las filas donde gmap_id está duplicado
filas_duplicadas = df1[df1['gmap_id'].duplicated(keep=False)]


In [ ]:
# Ordenar el DataFrame por la columna gmap_id
df_ordenado = filas_duplicadas.sort_values(by='gmap_id')


df_ordenado


In [ ]:
df1.iloc[:,[0,1,2,3,14]]

name  \
0                       Porter Pharmacy   
1                          City Textile   
2                          San Soo Dang   
3                          Nova Fabrics   
4                      Nobel Textile Co   
...                                 ...   
274996                     Maven Realty   
274997               Stone Castle Fence   
274998  Health Insurance Solutions Team   
274999          Bernard Building Center   
275000    HMR Motor Works and Auto Care   

                                                                       address  \
0                          Porter Pharmacy, 129 N Second St, Cochran, GA 31014   
1                        City Textile, 3001 E Pico Blvd, Los Angeles, CA 90023   
2                       San Soo Dang, 761 S Vermont Ave, Los Angeles, CA 90005   
3                          Nova Fabrics, 2200 E 11th St, Los Angeles, CA 90021   
4                        Nobel Textile Co, 719 E 9th St, Los Angeles, CA 90021   
...                                                                        ...   
274996      Maven Realty, 2180 McMillan Ave #71672, North Charleston, SC 29405   
274997     Stone Castle Fence, 124 Cypress Gardens Rd, Moncks Corner, SC 29461   
274998  Health Insurance Solutions Team, 6650 Rivers Ave, Charleston, SC 29406   
274999            Bernard Building Center, 27300 M-32, M-32, Hillman, MI 49746   
275000                            HMR Motor Works and Auto Care, Mio, MI 48647   

                                      gmap_id description  \
0       0x88f16e41928ff687:0x883dad4fd048e8f8        None   
1       0x80c2c98c0e3c16fd:0x29ec8a728764fdf9        None   
2       0x80c2c778e3b73d33:0xbdc58662a4a97d49        None   
3        0x80c2c89923b27a41:0x32041559418d447        None   
4       0x80c2c632f933b073:0xc31785961fe826a6        None   
...                                       ...         ...   
274996  0x88fe64e7daceaaab:0x101c046dcadcc9c3        None   
274997  0x88fe59f4602c95fd:0x57e2b1904d9e6949        None   
274998  0x88fe71f9617c15e1:0x89518c1d251706bf        None   
274999  0x4d34c7517d6e6e29:0x7fe553ee2f090cf0        None   
275000  0x88208f4360c053ab:0xe3d6883f19601a01        None   

                                                                                                                     url  
0       https://www.google.com/maps/place//data=!4m2!3m1!1s0x88f16e41928ff687:0x883dad4fd048e8f8?authuser=-1&hl=en&gl=us  
1       https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c98c0e3c16fd:0x29ec8a728764fdf9?authuser=-1&hl=en&gl=us  
2       https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c778e3b73d33:0xbdc58662a4a97d49?authuser=-1&hl=en&gl=us  
3        https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c89923b27a41:0x32041559418d447?authuser=-1&hl=en&gl=us  
4       https://www.google.com/maps/place//data=!4m2!3m1!1s0x80c2c632f933b073:0xc31785961fe826a6?authuser=-1&hl=en&gl=us  
...                                                                                                                  ...  
274996  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88fe64e7daceaaab:0x101c046dcadcc9c3?authuser=-1&hl=en&gl=us  
274997  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88fe59f4602c95fd:0x57e2b1904d9e6949?authuser=-1&hl=en&gl=us  
274998  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88fe71f9617c15e1:0x89518c1d251706bf?authuser=-1&hl=en&gl=us  
274999  https://www.google.com/maps/place//data=!4m2!3m1!1s0x4d34c7517d6e6e29:0x7fe553ee2f090cf0?authuser=-1&hl=en&gl=us  
275000  https://www.google.com/maps/place//data=!4m2!3m1!1s0x88208f4360c053ab:0xe3d6883f19601a01?authuser=-1&hl=en&gl=us  

[275001 rows x 5 columns]

In [ ]:
# Crear un nuevo DataFrame usando loc
dfYeplp_copia = dfYeplp.copy()

Eliminar columnas duplicadas, desde la 14 a 27

In [ ]:
# Crear copia donde solo se excluyen las duplicadas.
dfYeplp_copia = dfYeplp_copia.iloc[:,0:14]

In [ ]:
dfYeplp_copia.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
dtypes: object(14)
memory usage: 17.2+ MB


Ranking de ciudades con mas comentarios segun Yelp

In [ ]:
rankingFrecuenciaCiudad(dfYeplp_copia, 5)

NameError: name 'rankingFrecuenciaCiudad' is not defined

In [ ]:
# Eliminar las columnas 'url', 'relative_results', 'state' y 'price'
df_cleaned = df1.drop(columns=['url', 'relative_results', 'state', 'price'])

# Ver las primeras filas del DataFrame limpio
print(df_cleaned.head())

               name                                            address  \
0   Porter Pharmacy  Porter Pharmacy, 129 N Second St, Cochran, GA ...   
1      City Textile  City Textile, 3001 E Pico Blvd, Los Angeles, C...   
2      San Soo Dang  San Soo Dang, 761 S Vermont Ave, Los Angeles, ...   
3      Nova Fabrics  Nova Fabrics, 2200 E 11th St, Los Angeles, CA ...   
4  Nobel Textile Co  Nobel Textile Co, 719 E 9th St, Los Angeles, C...   

                                 gmap_id description   latitude   longitude  \
0  0x88f16e41928ff687:0x883dad4fd048e8f8        None  32.388300  -83.357100   
1  0x80c2c98c0e3c16fd:0x29ec8a728764fdf9        None  34.018891 -118.215290   
2  0x80c2c778e3b73d33:0xbdc58662a4a97d49        None  34.058092 -118.292130   
3   0x80c2c89923b27a41:0x32041559418d447        None  34.023669 -118.232930   
4  0x80c2c632f933b073:0xc31785961fe826a6        None  34.036694 -118.249421   

              category  avg_rating  num_of_reviews  \
0           [Pharmacy]    

In [ ]:
print(df_cleaned['MISC'][0])


{'Service options': ['In-store shopping', 'Same-day delivery'], 'Health & safety': ['Mask required', 'Staff required to disinfect surfaces between visits'], 'Accessibility': ['Wheelchair accessible entrance'], 'Planning': ['Quick visit']}


In [ ]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 275001 entries, 0 to 275000
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   name            274994 non-null  object 
 1   address         264939 non-null  object 
 2   gmap_id         275001 non-null  object 
 3   description     13155 non-null   object 
 4   latitude        275001 non-null  float64
 5   longitude       275001 non-null  float64
 6   category        272740 non-null  object 
 7   avg_rating      275001 non-null  float64
 8   num_of_reviews  275001 non-null  int64  
 9   hours           192448 non-null  object 
 10  MISC            194972 non-null  object 
dtypes: float64(3), int64(1), object(7)
memory usage: 23.1+ MB


--------------------------------------------------------------------------------

--------------------------------------------------------------------------------

#  **REVIEWS** YELP

copy paste de la funcion de arriba con la unica diferencia que se separo la funcion de contar los valores duplicados para mejor legibilidad, ya que en este caso se van a estar procesando varios archivos de review

In [7]:
def nulos_vacios_yelp_reviews(df):
    """
    Este procedimiento recibe un DataFrame y muestra la cantidad de valores nulos, vacíos y la cantidad de
    valores duplicados en todas las columnas excepto en las 6, 10, 11 y 13 del DataFrame. También muestra
    el mínimo, máximo y media para columnas de tipo float e integer.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    print("Informe de Nulos, Vacíos y Duplicados:")
    print("-" * 30)

    # Total de registros en el DataFrame
    total_registros = len(df)

    # Imprimir total de registros
    print(f"Total de registros: {total_registros}")
    print("-" * 30)

    # Iterar sobre todas las columnas para mostrar nulos y vacíos
    for columna in df.columns:
        # Contar valores nulos
        nulos = df[columna].isnull().sum()

        # Contar valores vacíos (considerando que solo se cuentan si son cadenas)
        vacios = (df[columna] == '').sum() if df[columna].dtype == 'object' else 0

        # Calcular porcentajes
        porcentaje_nulos = (nulos / total_registros) * 100 if total_registros > 0 else 0
        porcentaje_vacios = (vacios / total_registros) * 100 if total_registros > 0 else 0

        # Imprimir resultados
        print(f"Columna: {columna}")
        print(f"  Valores Nulos: {nulos}")
        print(f"  Valores Vacíos: {vacios}")
        print(f"  Porcentaje Nulos: {porcentaje_nulos:.2f}%")
        print(f"  Porcentaje Vacíos: {porcentaje_vacios:.2f}%")

        # Contar y mostrar la cantidad de valores duplicados solo si no es una de las columnas con datos anidados
        # Mostrar mínimo, máximo y media si la columna es de tipo float o int
        if df[columna].dtype in ['float64', 'int64']:
            minimo = df[columna].min()
            maximo = df[columna].max()
            media = df[columna].mean()
            print(f"  Mínimo: {minimo}")
            print(f"  Máximo: {maximo}")
            print(f"  Media: {media:.2f}")
        print("-" * 30)

def duplicados_yelp_reviews(df):
    # Inicializar un diccionario para almacenar los resultados
    resultados = {}

    for columna in df.columns:
        # Revisar si la columna es una de las que queremos analizar
        if columna in [df.columns[7], df.columns[0]]:
            # Filtrar los valores no nulos antes de contar duplicados
            cantidad_duplicados = df[columna].dropna().duplicated(keep=False).sum()
            resultados[columna] = cantidad_duplicados
            print(f"Cantidad de valores duplicados en {columna} (solo no nulos): {cantidad_duplicados}")

    # Devolver el diccionario con los resultados
    return resultados

# se realiza la revision de los archivos de review0,review1, [...], review6

In [ ]:
archivos = 0

while archivos < 7:#dado que son 7 archivos de review
    """
    aca vamos a recorrer archivo por archivo desde el review0 hasta el review6
    en la busqueda de datos vacios o nulos
    """
    #
    ruta = f'/content/drive/MyDrive/Proyecto_final/data/review{archivos}.parquet'
    print("________"*60)
    print("\n"*2)
    print(f"< archivo :  review{archivos}.parquet >")
    df = pd.read_parquet(ruta)
    nulos_vacios_yelp_reviews(df)
    archivos += 1

________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________
< archivo :  review0.parquet >
Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 1000000
------------------------------
Columna: review_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: user_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: business_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulo

**Como se puede ver en la data, no hay valores nulos o faltantes en cada uno de los archivos revisados de reviews, todas las columnas dan 0 valores faltantes o strings vacios**

Lo que se puede revisar ahora es:
- si hay datos duplicados, mas precisamente si hay review_id duplicados o textos duplicados, ya que seria raro encontrar dos comentarios identicos de dos personas distintas

In [ ]:
archivos = 0

while archivos < 7:#dado que son 7 archivos de review
    """
    aca vamos a recorrer archivo por archivo desde el review0 hasta el review6
    en la busqueda de datos duplicados
    """
    #aca vamos a recorrer archivo por archivo desde el review0 hasta el review6
    ruta = f'/content/drive/MyDrive/Proyecto_final/data/review{archivos}.parquet'
    print("________"*60)
    print("\n")
    print(f"< archivo :  review{archivos}.parquet >")
    df = pd.read_parquet(ruta)
    duplicados_yelp_reviews(df)
    archivos += 1

________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________


< archivo :  review0.parquet >
Cantidad de valores duplicados en review_id (solo no nulos): 0
Cantidad de valores duplicados en text (solo no nulos): 3247
__________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________

chequeamos si los duplicados no son mensajes cortos, como 'good service', 'i don't like', etc. o algun otro mensaje generico

In [ ]:
df['text'].value_counts()

text
Still waiting for a call back from the manager or owner......\nSpoke with manager, she had the stylist call me.....\nIf I wanted to deal with her again I would have\nIt's been weeks now.\nTerrible haircut, the worst quality tape in extensions on top of the fact that they were not washed to get the lament coating off before stylist dyed them. Did I mention she got me blonde to match my almost black hair. When questioned, she stated because I had purple added to my hair color the day before. Then why not add purple to dark brown? No answer, color never stayed. \nCompletely washed out, let stylist attempt to re do hair several times and just got worse. Installation is poor at best even though she claimed to be trained in thin hair and tape in extensions.....\nTook twice the time in the morning to try to hide the tape on extension because they were so close to hair line and the crown of my head.\nI wear them because I'm balding from a stressful event in my life......\nIn the end I've gone to drop them off to the stylist 3 times and she is alway gone getting food or not in.\nHer solution is to get them replaced by the company. I've already had to buy new hair thank you. And how can you get them replaced when it your screw up? Plus I don't want to let you touch my head again....\nSo small claims court here we come....\nPrice of hair, all the cuts you charged me for and the useless products you sold me to keep the color in the mess you created.\nI think 3+ months trying to resolve this is more than fair.\nLesson, don't offer services you don't have properly trained to preform...                                                                                                                                                                       4
They moved us 5 years ago about 4 miles. We were  ssatisfied except for the insert for our coffee table that they broke but didn't tell us. We found the parts in the driveway next day. This time we moved from the 2nd floor condo to the 4th floor at a cost of over $700.00 dollars. They were good guys that worked hard. They were about done when one of them came to me and said we had a problem. He said , his words!! I put a small hole in your water bed. I told him i don't even know how to answer that. He showed me the matress  laying on the floor. Water all over the floor. The matress that we slept on the night before was still sitting on the bed after being drained. No water on the bed. They loaded it on a furniture dolly after putting tape on the hole. I told them I could not patch it due to having 3 floors under me. I ask them to take it with them but was told that was not allowed. This is the same bed moved 5 years earlier.we called Mike and was told their was nothing he could do. He would get in touch with his boss Ryan. My wife spoke to Ryan today and was told he spoke to his men and believed them. They say now the hole was already there. How did we sleep on a water matress without a leak the night before. So he is believing them over us. To me that's the same as calling us lires. We were not asking for a new matress but he should have offered something. We even said just give us back the 50 dollar tip. NO. My wife is a christian and would not lie if you set her on fire. I will be talking to a lawyer. I would not deal with a company with low standard's like this 2 men and a truck. They didn't even need the truck. Ryan is a lier and not to be trusted. Stay away from them. I will write again after I speak with my attorney. Terry B    4
One of my favorite places to eat.  They have the best nachos in town. Their fruit tea is pretty good too.                                                                                                                                                                                                                                                                                                                                                                    

##  

---
##  
 - *Dado que son mensajes elaborados lo que vamos a hacer es realizar una comparacion de estos datos para verificar que la fila sea identica*

In [ ]:
archivos = 0
lista_duplicado = [] #almacenamos aca los indices que nos va a devolver la funcion duplicated().index.tolist()
while archivos < 7:#dado que son 7 archivos de review
    ruta = f'/content/drive/MyDrive/Proyecto_final/data/review{archivos}.parquet'
    print(f"< archivo :  review{archivos}.parquet >") #simplemente para verificar de que archivo va cargando el proceso
    df = pd.read_parquet(ruta)
    indices_duplicados = df[df['text'].duplicated(keep=False)].index.tolist() # guardamos los indices duplicados
    lista_duplicado.append(indices_duplicados) # los agregamos a la lista
    archivos += 1

< archivo :  review0.parquet >
< archivo :  review1.parquet >
< archivo :  review2.parquet >
< archivo :  review3.parquet >
< archivo :  review4.parquet >
< archivo :  review5.parquet >
< archivo :  review6.parquet >


##  
solo para chequear que hayan valores....

In [ ]:
lista_duplicado[0][0:5]

[749, 1283, 2244, 2855, 3124]

se revisa algunas filas para corroborar si los datos de las otras columnas coinciden en su mayoria

In [ ]:
review0 = pd.read_parquet('/content/drive/MyDrive/Proyecto_final/data/review0.parquet')
duplicados = review0.iloc[lista_duplicado[0]]

Vamos a elegir algunos datos al azar

In [ ]:
duplicados

review_id                 user_id  \
749     yCNf7tWAahLZVULW9dEhMw  BdWc0IO_lEjV7FlKv_4oZg   
1283    Aj0UVTXgRLXAasN1X4qXuA  jcgl3ygD81CP-runH1UYRg   
2244    RuPu2dWG2WZIzEV7nSx-qg  MrqeVVEux5nIO3Q76fZFHg   
2855    xgeOvTb78ZxUqrWH1limeA  uSji70QpgHIydaUzigMCgQ   
3124    4F7-40UML8B32d7ApJ76gw  Nwq0AyciH1n9E9Z3AyasLg   
...                        ...                     ...   
993853  SxrEHXbW5FKZCxwJzBAtKg  8uEChvHTHK395lShY2rY3g   
997789  ZR0oJAEW3OAT209ZrypdcQ  DUM2REPlxY8oDHN14x0yow   
997960  TFq85YwKoe-k-Ai8KEgegw  SyABteXA7u_-GA2nKLJ-XQ   
998451  OyG-kVjIHP858iB1OQNNlQ  obMfEZH4L3fjWJBqe2VArg   
998847  Ww8qU3cwb0DFqLFNzIOMWg  FS3vkaoquDXSDoUtT_oJJQ   

                   business_id  stars  useful  funny  cool  \
749     6N9T23_iXkcOCh6boAXpIg      4       0      0     0   
1283    gKPWC7F219oEbu5NXPzH4g      5       0      0     0   
2244    syuKX2pH3JJhOOppzNI6SA      5       1      0     0   
2855    AWnzFqIrlkLAKTc46_zmEg      5       0      0     0   
3124    J5-Y4ujpElrzPGJLYJ4tEg      3       0      0     0   
...                        ...    ...     ...    ...   ...   
993853  qMYCEl7r1K6WYmJglKaryA      4       0      0     0   
997789  TSKCK51GymAsxiA-kx-RiA      2       0      0     0   
997960  Nq1Kn7ud84upBHz9chFW0A      4       1      0     0   
998451  EF57en2jZd9e0m3E-TKV6Q      5       5      1     2   
998847  452karCazKBD2qXw6pLU-Q      5       2      0     0   

                                                     text                date  
749     I've always had good experiences here. The foo... 2015-08-30 19:50:43  
1283    Great food... Love the Fish tacos & Nachos are... 2016-05-07 10:14:52  
2244    We were back last night for the second time: a... 2018-08-04 23:52:52  
2855    I ate like a Queen!  A breakfeast Queen- and I... 2015-06-21 00:05:04  
3124    This is the best pizza by far in Reno. Crisp c... 2011-09-26 16:36:22  
...                                                   ...                 ...  
993853  Came out today to celebrate a friends birthday... 2017-10-20 21:25:50  
997789  Great atmosphere....a beautiful space. Ate din... 2015-05-24 16:29:05  
997960  Yummy sandwiches, live dj on Friday night. Gre... 2016-10-21 19:13:12  
998451  Exceeded my expectations for the price. Staff ... 2015-06-29 20:02:27  
998847  I did a long distance move from Tampa to NJ, t... 2016-07-02 12:26:44  

[3247 rows x 9 columns]

In [ ]:
duplicado = review0[review0['text'].str.contains("Great food... Love the Fish tacos & Nachos")]
duplicado

review_id                 user_id  \
1283    Aj0UVTXgRLXAasN1X4qXuA  jcgl3ygD81CP-runH1UYRg   
106898  5QxU40024gw-5lp9LZLTnQ  jcgl3ygD81CP-runH1UYRg   

                   business_id  stars  useful  funny  cool  \
1283    gKPWC7F219oEbu5NXPzH4g      5       0      0     0   
106898  gKPWC7F219oEbu5NXPzH4g      5       0      0     0   

                                                     text                date  
1283    Great food... Love the Fish tacos & Nachos are... 2016-05-07 10:14:52  
106898  Great food... Love the Fish tacos & Nachos are... 2016-03-21 03:52:16

In [ ]:
duplicado = review0[review0['text'].str.contains("Yummy sandwiches, live dj on Friday night")]
duplicado

review_id                 user_id  \
826111  n4oyGTnUWL7W_k3IHDZ0OA  SyABteXA7u_-GA2nKLJ-XQ   
997960  TFq85YwKoe-k-Ai8KEgegw  SyABteXA7u_-GA2nKLJ-XQ   
999063  620nZUdSFKK_yNLY8cM06g  SyABteXA7u_-GA2nKLJ-XQ   

                   business_id  stars  useful  funny  cool  \
826111  Nq1Kn7ud84upBHz9chFW0A      4       1      0     0   
997960  Nq1Kn7ud84upBHz9chFW0A      4       1      0     0   
999063  Nq1Kn7ud84upBHz9chFW0A      4       0      0     0   

                                                     text                date  
826111  Yummy sandwiches, live dj on Friday night. Gre... 2016-09-11 06:23:47  
997960  Yummy sandwiches, live dj on Friday night. Gre... 2016-10-21 19:13:12  
999063  Yummy sandwiches, live dj on Friday night. Gre... 2016-05-28 05:03:38

**Claramente los datos son casi identicos, usualmente lo unico que suele cambiar es la fecha y el review_id, por lo que podriamos decir que son datos que se ingresaron mal por algun bug de la aplicacion.**

#  

#  

#  **USERS** YELP

In [5]:
def duplicados_yelp(df,x:int):
  """
  esta aplicacion va a recorrer las columnas seleccionadas para verificar si hay datos duplicados
  """
  for columna in df.columns:
    #la idea es contar si hay review_id o textos duplicados en el df(cosa que no deberia haber)
    if columna in df.columns[x]: # solo vamos a observar la columna x del DF
      # Filtrar los valores no nulos antes de contar duplicados
      cantidad_duplicados = df[columna].dropna().duplicated(keep=False).sum()
      print(f"  Cantidad de valores duplicados en {columna} (solo no nulos): {cantidad_duplicados}")
      return cantidad_duplicados

In [8]:
archivos = 0

while archivos < 3:#dado que son 3 archivos lo seteamos para que recorra los 3 archivos de userX.parquet
    ruta = f'/content/drive/MyDrive/Proyecto_final/data/user{archivos}.parquet'
    print("________"*60)
    print("\n"*2)
    print(f"< archivo :  user{archivos}.parquet >")
    df = pd.read_parquet(ruta)
    nulos_vacios_yelp_reviews(df)
    archivos += 1

________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________



< archivo :  user0.parquet >
Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 800000
------------------------------
Columna: user_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: name
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: review_count
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0

**En este caso solo vamos a revisar la columna user_id, que no deberia tener valores duplicados**

In [6]:
archivos = 0

while archivos < 3:#dado que son 3 archivos lo seteamos para que recorra los 3 archivos de userX.parquet
    ruta = f'/content/drive/MyDrive/Proyecto_final/data/user{archivos}.parquet'
    print("________"*30)
    print("\n"*2)
    print(f"< archivo :  user{archivos}.parquet >")
    df = pd.read_parquet(ruta)
    df.reset_index(drop=True, inplace=True) #para reordenar el dataframe que aparecia como si hubiese sido utilizado por alguna funcion como concat, group by, u otra
    #print(df.head(2))
    duplicados_yelp(df,0) #ingresamos el df a analizar y la columna que vamos a chequear que no este duplicada
    archivos += 1

________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________



< archivo :  user0.parquet >
  Cantidad de valores duplicados en user_id (solo no nulos): 0
________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________



< archivo :  user1.parquet >
  Cantidad de valores duplicados en user_id (solo no nulos): 0
________________________________________________________________________________________________________________________________________________________________________________________________________________________________________________



< archivo :  user2.parquet >
  Cantidad de valores duplicados en user_id (solo no nu

chequeamos que la longitud coincida con el valor que nos dio la funcion

In [10]:
lista_indices_duplicados = []
lista_indices = df[df['user_id'].duplicated(keep=False)].index.tolist()
lista_indices_duplicados.append(lista_indices)
len(lista_indices_duplicados[0])

106600

verificamos algunos usuarios al azar

In [12]:
df_duplicados = df[df.index.isin(lista_indices_duplicados[0])]
df_duplicados.reset_index(drop=True, inplace=True)
df_duplicados

user_id       name  review_count        yelping_since  \
0       qVc8ODYU5SZjKXVBgXdI7w     Walker           585  2007-01-25 16:47:26   
1       j14WgRoU_-2ZE1aw1dXrJg     Daniel          4333  2009-01-25 04:35:42   
2       2WnXYQFK0hXEoTxPtV2zvg      Steph           665  2008-07-25 10:41:00   
3       SZDeASXq7o05mMNLshsdIA       Gwen           224  2005-11-29 04:38:33   
4       hA5lMy-EnncsH4JoR-hFGQ      Karen            79  2007-01-05 19:40:59   
...                        ...        ...           ...                  ...   
106595  Ls7UQtY3YhXP9CecCbx3cg     Mighty            10  2012-07-02 02:42:52   
106596  -lJvVg3jiFg6k9PsOOhwcg      Kiley            14  2011-01-29 05:41:42   
106597  nhR_swywxzzvgSCBpVCpVg       Kent             3  2009-01-31 04:35:28   
106598  G8UL6NO3kqnlhweKvoV3Dw  Kalimuthu            74  2011-07-03 02:45:01   
106599  EIpPME2cIuNbhfXrtsSApg     Monica           397  2010-06-12 15:33:02   

        useful  funny   cool  \
0         7217   1259   5994   
1        43091  13066  27281   
2         2086   1010   1003   
3          512    330    299   
4           29     15      7   
...        ...    ...    ...   
106595       4      1      0   
106596      21      2      0   
106597       7      0      0   
106598     118      8     57   
106599     785    155    403   

                                                    elite  \
0                                                    2007   
1       2009,2010,2011,2012,2013,2014,2015,2016,2017,2...   
2                                2009,2010,2011,2012,2013   
3                                          2009,2010,2011   
4                                                           
...                                                   ...   
106595                                                      
106596                                                      
106597                                                      
106598                                     2016,2017,2018   
106599           2014,2015,2016,2017,2018,2019,20,20,2021   

                                                  friends  fans  ...  \
0       NSCy54eWehBJyZdG2iE84w, pe42u7DcCH2QmI81NX-8qA...   267  ...   
1       ueRPE0CX75ePGMqOFVj6IQ, 52oH4DrRvzzl8wh5UXyU0A...  3138  ...   
2       LuO3Bn4f3rlhyHIaNfTlnA, j9B4XdHUhDfTKVecyWQgyA...    52  ...   
3       enx1vVPnfdNUdPho6PH_wg, 4wOcvMLtU6a9Lslggq74Vg...    28  ...   
4       PBK4q9KEEBHhFvSXCUirIw, 3FWPpM7KU1gXeOM_ZbYMbA...     1  ...   
...                                                   ...   ...  ...   
106595  aoLZdOkg7yVb8AbfupaYkA, NpZ4eXCC_EPub70NS7FdKA...     0  ...   
106596  yGqAQ7CSqLJcp_qAFwyS0w, QF-aFIZRB8fnLwJQAs9SbQ...     0  ...   
106597  GSAI2KsMK6qu0Eh624ze0A, 2-veVnvUPcQvCAj35Cp_Nw...     2  ...   
106598  M-dJGDgNpiJEhmzkJBK25Q, BTBUZS9HFrLntinRh9BsHw...     7  ...   
106599  sVr6_v-lwixu3VREOO9XpA, zxuxd6Hz2tKcpgZ71dYEcw...    38  ...   

        compliment_more  compliment_profile  compliment_cute  compliment_list  \
0                    65                  55               56               18   
1                   264                 184              157              251   
2                    13                  10               17                3   
3                     4                   1                6                2   
4                     1                   0                0                0   
...                 ...                 ...              ...              ...   
106595                0                   0                0                0   
106596                0                   0                0                0   
106597                0                   0                0                0   
106598                3                   0                0                0   
106599                3                   1                1                1   

        compliment_note  compliment_plain  compliment_cool  compliment_funny  \
0                   232   

In [13]:
usuario_random = df_duplicados[df_duplicados['user_id']=='qVc8ODYU5SZjKXVBgXdI7w']
usuario_random
#

user_id    name  review_count        yelping_since  \
0      qVc8ODYU5SZjKXVBgXdI7w  Walker           585  2007-01-25 16:47:26   
11100  qVc8ODYU5SZjKXVBgXdI7w  Walker           585  2007-01-25 16:47:26   

       useful  funny  cool elite  \
0        7217   1259  5994  2007   
11100    7217   1259  5994  2007   

                                                 friends  fans  ...  \
0      NSCy54eWehBJyZdG2iE84w, pe42u7DcCH2QmI81NX-8qA...   267  ...   
11100  NSCy54eWehBJyZdG2iE84w, pe42u7DcCH2QmI81NX-8qA...   267  ...   

       compliment_more  compliment_profile  compliment_cute  compliment_list  \
0                   65                  55               56               18   
11100               65                  55               56               18   

       compliment_note  compliment_plain  compliment_cool  compliment_funny  \
0                  232               844              467               467   
11100              232               844              467               467   

       compliment_writer  compliment_photos  
0                    239                180  
11100                239                180  

[2 rows x 22 columns]

In [14]:
usuario_random1 = df_duplicados[df_duplicados['user_id']=='-lJvVg3jiFg6k9PsOOhwcg']
usuario_random1

user_id   name  review_count        yelping_since  \
43296   -lJvVg3jiFg6k9PsOOhwcg  Kiley            14  2011-01-29 05:41:42   
64396   -lJvVg3jiFg6k9PsOOhwcg  Kiley            14  2011-01-29 05:41:42   
85496   -lJvVg3jiFg6k9PsOOhwcg  Kiley            14  2011-01-29 05:41:42   
106596  -lJvVg3jiFg6k9PsOOhwcg  Kiley            14  2011-01-29 05:41:42   

        useful  funny  cool elite  \
43296       21      2     0         
64396       21      2     0         
85496       21      2     0         
106596      21      2     0         

                                                  friends  fans  ...  \
43296   yGqAQ7CSqLJcp_qAFwyS0w, QF-aFIZRB8fnLwJQAs9SbQ...     0  ...   
64396   yGqAQ7CSqLJcp_qAFwyS0w, QF-aFIZRB8fnLwJQAs9SbQ...     0  ...   
85496   yGqAQ7CSqLJcp_qAFwyS0w, QF-aFIZRB8fnLwJQAs9SbQ...     0  ...   
106596  yGqAQ7CSqLJcp_qAFwyS0w, QF-aFIZRB8fnLwJQAs9SbQ...     0  ...   

        compliment_more  compliment_profile  compliment_cute  compliment_list  \
43296                 0                   0                0                0   
64396                 0                   0                0                0   
85496                 0                   0                0                0   
106596                0                   0                0                0   

        compliment_note  compliment_plain  compliment_cool  compliment_funny  \
43296                 0                 0                0                 0   
64396                 0                 0                0                 0   
85496                 0                 0                0                 0   
106596                0                 0                0                 0   

        compliment_writer  compliment_photos  
43296                   0                  0  
64396                   0                  0  
85496                   0                  0  
106596                  0                  0  

[4 rows x 22 columns]

In [15]:
usuario_random2 = df_duplicados[df_duplicados['user_id']=='SZDeASXq7o05mMNLshsdIA']
usuario_random2

user_id  name  review_count        yelping_since  \
3      SZDeASXq7o05mMNLshsdIA  Gwen           224  2005-11-29 04:38:33   
11103  SZDeASXq7o05mMNLshsdIA  Gwen           224  2005-11-29 04:38:33   

       useful  funny  cool           elite  \
3         512    330   299  2009,2010,2011   
11103     512    330   299  2009,2010,2011   

                                                 friends  fans  ...  \
3      enx1vVPnfdNUdPho6PH_wg, 4wOcvMLtU6a9Lslggq74Vg...    28  ...   
11103  enx1vVPnfdNUdPho6PH_wg, 4wOcvMLtU6a9Lslggq74Vg...    28  ...   

       compliment_more  compliment_profile  compliment_cute  compliment_list  \
3                    4                   1                6                2   
11103                4                   1                6                2   

       compliment_note  compliment_plain  compliment_cool  compliment_funny  \
3                   12                16               26                26   
11103               12                16               26                26   

       compliment_writer  compliment_photos  
3                     10                  9  
11103                 10                  9  

[2 rows x 22 columns]

In [31]:
checkin = pd.read_parquet('/content/drive/MyDrive/Proyecto_final/data/checkin.parquet')
checkin.head()

business_id                                               date
0  ---kPU91CF4Lq2-WlRu9Lw  2020-03-13 21:10:56, 2020-06-02 22:18:06, 2020...
1  --0iUa4sNDFiZFrAdIWhZQ  2010-09-13 21:43:09, 2011-05-04 23:08:15, 2011...
2  --30_8IhuyMHbSOcNWd6DQ           2013-06-14 23:29:17, 2014-08-13 23:20:22
3  --7PUidqRWpRSpXebiyxTg  2011-02-15 17:12:00, 2011-07-28 02:46:10, 2012...
4  --7jw19RH9JKXgFohspgQw  2014-04-21 20:42:11, 2014-04-28 21:04:46, 2014...

In [ ]:
nulos_vacios_yelp_reviews(checkin)

Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 131930
------------------------------
Columna: business_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: date
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------


In [37]:
duplicados_yelp(checkin,0)

  Cantidad de valores duplicados en business_id (solo no nulos): 0


0

In [32]:
duplicados_yelp(checkin,1)

  Cantidad de valores duplicados en date (solo no nulos): 0


0

In [ ]:
tip = pd.read_parquet('/content/drive/MyDrive/Proyecto_final/data/tip.parquet')
tip.head()

user_id             business_id  \
0  AGNUgVwnZUey3gcPCJ76iw  3uLgwr0qeCNMjKenHJwPGQ   
1  NBN4MgHP9D3cw--SnauTkA  QoezRbYQncpRqyrLH6Iqjg   
2  -copOvldyKh1qr-vzkDEvw  MYoRNLb5chwjQe3c_k37Gg   
3  FjMQVZjSqY8syIO-53KFKw  hV-bABTK-glh5wj31ps_Jw   
4  ld0AperBXk1h6UbqmM80zw  _uN0OudeJ3Zl_tf6nxg5ww   

                                                text                date  \
0                     Avengers time with the ladies. 2012-05-18 02:17:21   
1  They have lots of good deserts and tasty cuban... 2013-02-05 18:35:10   
2             It's open even when you think it isn't 2013-08-18 00:56:08   
3                          Very decent fried chicken 2017-06-27 23:05:38   
4             Appetizers.. platter special for lunch 2012-10-06 19:43:09   

   compliment_count  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

In [ ]:
nulos_vacios_yelp_reviews(tip)

Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 908915
------------------------------
Columna: user_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: business_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: text
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: date
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
------------------------------
Columna: compliment_count
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Mínimo: 0
  Máximo: 6
  Media: 0.01
------------------------------


In [13]:
duplicados_yelp(tip,2)

  Cantidad de valores duplicados en text (solo no nulos): 72992


72992

chequeamos los indices y verificamos que la longitud de la lista coincida con la cantidad de datos duplicados que nos da la funcion

In [18]:
indices_duplicados = tip[tip['text'].duplicated(keep=False)].index.tolist() # guardamos los indices duplicados
lista_duplicado = [] #almacenamos aca los indices que nos va a devolver la funcion duplicated().index.tolist()
lista_duplicado.append(indices_duplicados) # los agregamos a la lista
len(lista_duplicado[0])

72992

Ahora si usamos la mascara para filtrar los datos duplicados del df

In [20]:
duplicados = tip[tip.index.isin(lista_duplicado[0])]
duplicados.info()

<class 'pandas.core.frame.DataFrame'>
Index: 72992 entries, 8 to 908911
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   user_id           72992 non-null  object        
 1   business_id       72992 non-null  object        
 2   text              72992 non-null  object        
 3   date              72992 non-null  datetime64[ns]
 4   compliment_count  72992 non-null  int64         
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 3.3+ MB


#  

Vamos a elegir nuevamente algunos comentario al azar para corroborar si coinciden el resto de caracteristicas

In [21]:
duplicados

user_id             business_id  \
8       VL12EhEdT4OWqGq0nIqkzw  xODBZmX4EmlVvbqtKN7YKg   
40      q4twMHnYmRmKRxdVDYakbQ  u3Z7X2T7ssWO9qvMRNcsOA   
49      2eC4qmjUNR7-LG8s6t4Ehg  JQXKCQPIeoxzy6x119X6mQ   
55      rSeQlwsBuWzKZYKac7r8Gw  UFCN0bYdHroPKu6KV5CJqg   
59      2CCJdupLxjqxthneSuuScQ  tnJT4UUVsrrJQL7wfDe_lQ   
...                        ...                     ...   
908873  zCXrIBe0LhylkQW_1VA0SA  GXFMD0Z4jEVZBCsbPf4CTQ   
908874  ZPXSpWG59WKu127zrh6-TQ  O4mH_WvFA0My7qOxbkAYow   
908888  bHHTHTeM1Sb8-aun1pM8Tw  0OsR9lO16jxa0xWUY57s9g   
908907  2-vAo2UfkD7QHA5TG8kwmg  wQUBiBqlzC6cbdkX-GaBqQ   
908911  1uxtQAuJ2T5Xwa_wp7kUnA  OaGf0Dp56ARhQwIDT90w_g   

                                                     text                date  \
8                                                   Tacos 2012-07-27 01:48:24   
40                             Great place for breakfast! 2011-12-25 14:23:53   
49                                             Work lunch 2012-03-15 16:18:19   
55      Best lunch/ happy hour in downtown!!! Kenny ro... 2016-08-01 20:50:52   
59                                      Super Spicy Roll! 2010-12-24 16:53:36   
...                                                   ...                 ...   
908873                                          Excellent 2018-03-12 22:31:00   
908874                                       Great salad! 2018-07-08 19:27:04   
908888                                          Very good 2020-02-23 00:17:34   
908907                             The food was delicious 2021-03-12 00:15:07   
908911                            Great food and service. 2021-10-30 11:54:36   

        compliment_count  
8                      0  
40                     0  
49                     0  
55                     0  
59                     0  
...                  ...  
908873                 0  
908874                 0  
908888                 0  
908907                 0  
908911                 0  

[72992 rows x 5 columns]

In [23]:
datos_random = tip[tip['text']=="Tacos"]
datos_random

user_id             business_id   text  \
8       VL12EhEdT4OWqGq0nIqkzw  xODBZmX4EmlVvbqtKN7YKg  Tacos   
13341   kVqjvUueC7hWZ1_IjkGjvg  mFIffumYpySeZnQ3ckqpDA  Tacos   
284890  cuAHifRAkYY9iEVZkRlpXA  vhDWGF-8BfsxvS7Zo5Wv2w  Tacos   
373167  Rr4cLb6Go91FT134o6RsKg  qFwKNXKt2fvCatrLAqDONg  Tacos   
412667  HXB_ByCCp8cZm1WMjKn1Tw  SCjUeZfjafJ88fWWQjRh2w  Tacos   
425498  M5HoGdy-mNOV7EHhxB9P9A  HwyR8zWjIomCDrHAJ0cXSQ  Tacos   
833946  ccMnj_hT1QJfUuMrL9644g  HF9LX5776QYYK70tyIL8gw  Tacos   
906357  R53MDGKiur16lY7g_B74Eg  6gCQvovXf3wmrfwuTyeLlQ  Tacos   

                      date  compliment_count  
8      2012-07-27 01:48:24                 0  
13341  2010-02-20 16:54:36                 0  
284890 2018-07-29 04:40:56                 0  
373167 2012-05-13 01:47:47                 0  
412667 2011-08-26 21:35:35                 0  
425498 2013-05-24 03:24:25                 0  
833946 2013-01-28 22:03:37                 0  
906357 2012-02-16 18:45:07                 0

In [25]:
datos_random1 = tip[tip['text']=="Great food and service."]
datos_random1

user_id             business_id  \
24773   G52P5bsFM3AjN64MKpPS0Q  2fuQnVPwhU_zzfUog_JGsw   
52795   xWR-GSPPlt7taGuM1vk_0g  8kUh6TROemLfbVR_ewVVLg   
67037   wemLDAZ-sPdxlSt9VvkeNg  5oX4G1cptixPZecih9L3dg   
89924   xKVTd1u9A4ucHp__g4U-Ew  A5sXEcVHYnmNDs0bwb7u-w   
173407  SiXy8yxpjhZPcTeI0BjPeQ  s63cVEM7S0SSBQIkUIsfJQ   
...                        ...                     ...   
893407  ezuHFcqzu9PFOq2ajq1Geg  PzhPMkaNYiKDTHoTG0r8rw   
904078  rhAMh47RIlUIg0lyyGAXQQ  qt_E6txwQ1h62wyv8701UQ   
905776  ejo3D0lanSoG0PWDmHm03g  scoHdh0XDA_W7xAZW3zAKQ   
908754  EDicnxs2NPxktUc4KgHDdg  4A_1kNLFVYmQ-zfz93huBg   
908911  1uxtQAuJ2T5Xwa_wp7kUnA  OaGf0Dp56ARhQwIDT90w_g   

                           text                date  compliment_count  
24773   Great food and service. 2016-08-28 21:41:57                 0  
52795   Great food and service. 2015-10-24 00:47:48                 0  
67037   Great food and service. 2017-01-28 19:37:07                 0  
89924   Great food and service. 2018-01-27 11:59:31                 0  
173407  Great food and service. 2014-04-03 13:32:26                 0  
...                         ...                 ...               ...  
893407  Great food and service. 2016-04-30 23:22:42                 0  
904078  Great food and service. 2019-04-22 01:10:57                 0  
905776  Great food and service. 2021-06-26 20:22:11                 0  
908754  Great food and service. 2016-06-26 15:42:14                 0  
908911  Great food and service. 2021-10-30 11:54:36                 0  

[103 rows x 5 columns]

dado que son como sugerencias de diferentes lugares y diferentes usuarios, podemos dejarlo sin efecto
Lo que restaria revisar es si la caracteristica de conteo de cumplidos tiene mas datos que 0

In [29]:
tip.compliment_count.value_counts()

compliment_count
0    898376
1      9821
2       617
3        82
4        14
5         3
6         2
Name: count, dtype: int64